tests/malloc_thread_safety/main.c

```diff
@@ -78,41 +97,62 @@ int main(void)
-                       THREAD_CREATE_STACKTEST, t1_t2_func, NULL, "t2");
-    expect((t1 != KERNEL_PID_UNDEF) && (t2 != KERNEL_PID_UNDEF));
-
-    for (uint16_t i = 0; i < 2 * MS_PER_SEC; i++) {
-        xtimer_usleep(US_PER_MS);
-        /* shuffle t1 and t2 in their run queue. This should eventually hit
-         * during a call to malloc() or free() and disclose any missing
-         * guards */
-        evil_schedule_hack_dont_do_this_at_home(THREAD_PRIORITY_MAIN + 1);
-    }
-
-    /* Don't keep threads spinning */
-    atomic_store(&is_running, 0);
-    /* Give threads time to terminate */
-    xtimer_usleep(10 * US_PER_MS);
+    void *(* const funcs[])(void *) = { t1_t2_malloc_func, t1_t2_realloc_func };
+    const char *tests[] = { "malloc()/free()", "realloc()/free()" };
+
+    for (size_t i = 0; i < ARRAY_SIZE(funcs); i++) {
+        printf("Testing: %s\n", tests[i]);
+        t1 = thread_create(t1_stack, sizeof(t1_stack), THREAD_PRIORITY_MAIN + 1,
+                           THREAD_CREATE_STACKTEST, funcs[i], NULL, "t1");
+        t2 = thread_create(t2_stack, sizeof(t2_stack), THREAD_PRIORITY_MAIN + 1,
+                           THREAD_CREATE_STACKTEST, funcs[i], NULL, "t2");
+        expect((t1 != KERNEL_PID_UNDEF) && (t2 != KERNEL_PID_UNDEF));
+
+        for (uint16_t i = 0; i < 2 * MS_PER_SEC; i++) {
+            xtimer_usleep(US_PER_MS);
+            /* shuffle t1 and t2 in their run queue. This should eventually hit
+             * during a call to malloc() or free() and disclose any missing
+             * guards */
+            evil_schedule_hack_dont_do_this_at_home(THREAD_PRIORITY_MAIN + 1);
+        }
+
+        /* Don't keep threads spinning */
+        atomic_store(&is_running, 0);
+        /* Give threads time to terminate */
+        xtimer_usleep(10 * US_PER_MS);
 
 #ifndef NO_MALLINFO
-    struct mallinfo post = mallinfo();
-
-    /* RIOT's board or arch support hopefully doesn't use malloc, so there
-     * should be zero bytes allocated prior to the first call to malloc() in
-     * this test. But let's be forgiving and just expect that the number of
-     * allocated bytes before and after the test is equal.
-     */
-    expect(pre.uordblks == post.uordblks);
+        struct mallinfo post = mallinfo();
+
+        /* RIOT's board or arch support hopefully doesn't use malloc, so there
+         * should be zero bytes allocated prior to the first call to malloc() in
+         * this test. But let's be forgiving and just expect that the number of
+         * allocated bytes before and after the test is equal.
+         */
+        if (pre.uordblks != post.uordblks) {
+            failed = 1;
+            puts("Not all blocks were correctly freed!");
+            printf("mallinfo().uordblks before test: %u, after test: %u\n",
+                   (unsigned)pre.uordblks, (unsigned)post.uordblks);
+        }
 #endif
-    puts("TEST PASSED");
+    }
+
+    if (failed) {
```

Comment:
If `ifndef NO_MALLINFO` this can never be false, does this global check make sense in that case? What is considered a passing test in that case?

Reply:
The actual test is that this doesn't crash. E.g. without the mutex I get hard faults or `malloc()` never returning. The additional `failed` only checks that `malloc()` and friends don't leak memory.

Maybe I should drop this, since this would only detect bugs in the standard c lib rather than in RIOT. On the other hand, this check is pretty much for free and it might be useful to detect a broken toolchain. There also was some discussion on an alternative implementation in the context of the Google summer of code, which would profit from this test as well.

Reply:
Makes sense, lets keep it mind adding a comment and suppressing the warning?